java.base/src/main/java/java/lang/Object.java

```diff
@@ -50,21 +58,61 @@ public boolean equals(Object other) {
         return this == other;
     }
 
-    public final native void notify();
+    public final void notify() {
+        getMonitor().signal();
+    }
 
-    public final native void notifyAll();
+    public final void notifyAll() {
+        getMonitor().signalAll();
+    }
 
     public String toString() {
         return getClass().getName() + "@" + Integer.toHexString(hashCode());
     }
 
     public final void wait() throws InterruptedException {
-        wait(0L);
+        getMonitor().await();
+    }
+
+    public final void wait(long millis) throws InterruptedException {
+        getMonitor().await(millis);
+    }
+
+    public final void wait(long millis, int nanos) throws InterruptedException {
+        getMonitor().await(millis, nanos);
     }
 
-    public final native void wait(long millis) throws InterruptedException;
+    @NoReflect
+    @Hidden
+    private void monitorEnter() {
+        getMonitor().enter();
+    }
+
+    @NoReflect
+    @Hidden
+    private void monitorExit() {
+        getMonitor().exit();
+    }
+
+    @NoReflect
+    @Hidden
+    private boolean holdsLock() {
+        Monitor monitor = addr_of(refToPtr(this).sel().monitor).loadSingleAcquire();
+        return monitor != null && monitor.isHeldByCurrentThread();
+    }
 
-    public final native void wait(long millis, int nanos) throws InterruptedException;
+    @NoReflect
+    private Monitor getMonitor() {
+        Monitor monitor = addr_of(refToPtr(this).sel().monitor).loadSingleAcquire();
+        if (monitor == null) {
+            monitor = new Monitor();
+            Monitor appearing = addr_of(refToPtr(this).sel().monitor).compareAndSwapRelease(null, monitor);
+            if (appearing != null) {
+                monitor = appearing;
+            }
+        }
+        return monitor;
+    }
 
     protected Object clone() throws CloneNotSupportedException {
         Class<?> clazz = this.getClass();
```

Comment:
Should we add code to `clone` to null out the monitor field of the freshly cloned object before it returns it?  That seems like the least magical/most robust way to do handle that

Reply:
Hmm yes good point...